### DISCUSSION TITLE
✅ COMPLETED - RfC - managing URLs in Django CMS

### DISCUSSION THREAD

**Discussion by fsbraun** [django-cms/django-cms#8024] [General]
Currently, there is no unified way of managing URLs with django CMS:

* **djangocms-link** contains a link plugin which (only) allows linking `Page` objects. It does not allow linking other models.
* **djangocms-frontend** contains a link plugin which allows other models to be configured to be linked, too. It also supports djangocms-url-manager (see below).
* **djngocms-text** contains a dynamic link feature which allows linking Pages.
* **djangocms-url-manager** is an attempt to centralize link management for django CMS 4, but as far as I know, has so far not generated much traction, potentially its additional level of abstraction which requires every single permissible URL to be managed in the admin.

IMHO minimal requirements for a URL managing apps are:
* Provide a text-enabled plugin for links.
* Provide an admin form field and widget for re-use in other apps, the widget should query the available link destinations dynamically
* Provide an admin endpoint for querying for available link targets for use by other apps (e.g., by djangocms-text for its dynamic link feature)
* Provide an interface based on v4's `cms_config` for other apps to register link targets or get the form fields/admin endpoint
* Provide some sort of autoconfig to get link targets from legacy or non-CMS apps.

Given the multiple places where solutions have been implemented, the question is: Where should this functionality live?

**I'd like to propose giving djangocms-link a boost and implementing the minimal requirements listed above.** Implementation should provide the new features for v4 apps through `cms_config` but keep v3 compatibility. 

**djangocms-url-manager can be the place for a more detailed management of links** - if there is a need for that. 

All other packages could then reuse these features.

Please add your thoughts, ideas or comments!

**Comment by macolo** [comment]
Hi @fsbraun thank you for this important point. 

I agree with leading with djangocms-link. 

I have two additional comments:

**1) extending inline linking in djangocms-text**
I wonder if it makes sense to extend the inline (not in-text) linking feature to allow to select not only pages but also other objects with an url? The more URLs, the harder it gets to choose the right URL from a searchable dropdown? Maybe instead of a searchable dropdown, a modal overlay would provide a more stable user experience for enterprise-size projects with potentially thousands of different URLs and multiple object types?


**2) Definition of Use Cases**
Here is how I would expect a developer to use a LinkField() and a djangocms-text HtmlField():

1) LinkField(): A developer wants to add a `LinkField()` to their custom plugin, for example a `TeaserPlugin` that features one single page, or a ButtonPlugin that renders a link as a custom styled button. Through the `LinkField()`, the dev could make all globally enabled objects available, or could configure specific ones to be available.

2) djangocms-text HtmlField() or RichTextField(): A developer creates a `HeroPlugin` by adding a `HtmlField or RichTextField` as "Lead Text" and another one as "Body text". Inside such fields (i.e. in Tip Tap), the developer wants editors to be able to link arbitrary (globally enabled) objects inline (i.e. not through the in-text plugins but directly in Tip Tap), including pages.

**Comment by marksweb** [comment]
> * **djangocms-url-manager** is an attempt to centralize link management for django CMS 4, but as far as I know, has so far not generated much traction, potentially its additional level of abstraction which requires every single permissible URL to be managed in the admin.

And this was part of the very specific usecase that v4 was developed with. I tried to start a project with this, but it's so different to what we're used to [coming from v3] that it's difficult without the docs/examples of how you use it.

**Reply by fsbraun** [reply]
I made the same experience.

**Comment by fsbraun** [comment]
I've put some work into a link widget which would be reusable in plugins (or all of admin, for that matter). It merges essentially the current `internal_link`, `external_link`, `file_link`, `anchor`, `mailto`, `phone`  fields - distributed over 2 fieldsets - into one field. I hope for an UX improvement.

It uses django admin's select2. 

![link-widget](https://github.com/user-attachments/assets/b3b921c4-f2b4-4420-8e19-97de3a6172a8)

~~(Internal links might need an additional site selector.)~~

**Reply by fsbraun** [reply]
Here's the PR: https://github.com/django-cms/djangocms-link/pull/229

**Comment by fsbraun** [comment]
djangocms-link 5.0.0 released 🎉
